### PR TITLE
enhancement(vrl): Add `tally` function

### PR DESCRIPTION
**Pull request by jszwedko**
This is an attempt to offer a possible solution for #8976

The name was inspired by Ruby: https://rubydoc.info/stdlib/core/Enumerable:tally

Pushing this up for discussion as this suffers from the same issue we've
been discussing around `parse_ruby_hash` in that our `vrl::Value` only
allows `String`s for keys, but in this case the elements could be
non-string.

Possible options:

1. Throw a runtime error if the array has non-string elements in it.
2. Add a mechanism to assert the type of all array elements. This would
  be the same as the first option, but allow us to fail at compile time.
3. Use some serialization format of `vrl::Value` as the keys
4. Update `vrl::Value::Map` to be a `BTreeMap<Value, Value>`. However
  we'd still have to identify how to map this back to `vector::Event`
  which requires `String` keys or error on assignment. We could update
  `vector::Value::Map` to also allow `vector::Value` keys; however, those keys
  would need to be serialized when serializing as, for example, JSON
  however, which only allows String keys.

I'm inclined towards the option 3 as the easiest solution that should
cover 90% of people's use-cases.

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

### CODE REVIEW

**Comment by StephenWakely** [comment]
I think 3 is the only realistic solution as unpleasant as it may be. 1 and 2 limit the usefulness of the tally function and 4 means we would have to convert the key using some serialization in order to copy the value into Vector. It also would not play nicely with the paths for the lookup.

I would vote for a json serialization (for all keys other than string) since it is familiar to most users. Things like this should then be possible, albeit unwieldly (did we fix escaping `"` in paths?):

```coffee
.thing = tally([{"greetings": 3}, {"greetings": 4}, {"greetings": 3}])
.greeted_three_times = .thing."{/"greetings/": 3}"
```

If we enable dynamic paths you could perhaps also do this:

```coffee
.greeted_three_times = .thing[encode_json({"greetings": 3})]
```

**Comment by JeanMertz** [comment]
I can't say I'm a big fan of either option on this list.

I wonder if we can punt on this (and, _if_ we want to change this, I'd propose an RFC for it), and use a different output format for this function?

That is:

```json
["foo", "foo", "bar", true]
```

```coffee
tally(.)
```

```json
[
	{ "value": "foo", "count": 2 },
	{ "value": "bar", "count": 1 },
	{ "value": true, "count": 1 },
]
```

**Comment by StephenWakely** [comment]
> Although, it would still require iteration to be useful:

It's probably worth noting that the entire `tally` function can be deprecated once iteration, specifically some form of `fold` function, is implemented. I believe `tally` is just a temporary measure to help a specific user move forward.

**Comment by JeanMertz** [comment]
> It's probably worth noting that the entire `tally` function can be deprecated once iteration, specifically some form of `fold` function, is implemented. I believe `tally` is just a temporary measure to help a specific user move forward.

It is, but we also plan on adding more "useful utility functions" to VRL, even if they could also be done manually with iteration. I'm not confident if this passes that threshold (it might be worth defining that threshold in the VRL design document), but it could.

**Comment by jszwedko** [comment]
> It's probably worth noting that the entire `tally` function can be deprecated once iteration, specifically some form of `fold` function, is implemented. I believe `tally` is just a temporary measure to help a specific user move forward.

Well, yes, they could use `fold` but they would run into the same issue I'm running into where they need to have string keys for the map they are build 😄 

Returning an array of key/values is interesting, but not very easy to use to lookup specific values from.

It would be nice to be able to enforce the type of all elements in an array at compile time. This, combined with iteration to map the types, thus letting the user decide how to marshal them, would probably be my preference.

**Comment by JeanMertz** [comment]
I verified that string-values is all we need for now, until iteration support lands.

So, I'd propose we add this function, using option (1) to error if a non-string value is passed in (making this function fallible). Additionally, we ship this function without documenting it so that we can back out of it (or change it) once iteration lands.

**Comment by jszwedko** [comment]
@StephenWakely @JeanMertz I updated this to only take string values if you could take a look.

Just noting again we decided not to document this function for now because it will likely go away when iteration support lands.

File: website/cue/reference/remap/functions/tally.cue [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package metadata
+
+remap: functions: unique: {
```
Comment (StephenWakely) [inline website/cue/reference/remap/functions/tally.cue:3]: I think this file shouldn't be here.

**Review by StephenWakely — APPROVED** [review]
Looks good to me apart from that stray cue file.

File: website/cue/reference/remap/functions/tally.cue (hunk 1 above)
Comment (jszwedko) [reply website/cue/reference/remap/functions/tally.cue:3]: Oops! Thank you 😄

**Review by JeanMertz — APPROVED** [review]
sweet! 🙌

File: lib/vrl/stdlib/src/tally.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            title: "tally",
+            source: r#"tally!(["foo", "bar", "foo", "baz"])"#,
+            result: Ok(r#"{"foo": 2, "bar": 1, "baz": 1]"#),
+        }]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::Compiler,
+        _info: &FunctionCompileContext,
```
[the remark this answers is not in the document]
Comment (jszwedko) [reply lib/vrl/stdlib/src/tally.rs]: 👍 updated in c1a2676b4

**Pull request merged**